collector_test.go

```diff
@@ -65,24 +212,28 @@ func TestPduToSample(t *testing.T) {
 				Oid:  "1.1.1.1.1",
 				Help: "Help string",
 			},
-			oidToPdu:           make(map[string]gosnmp.SnmpPDU),
-			expectedMetric:     `label:<name:"test_metric" value:"-2" > gauge:<value:1 > `,
-			expectedMetricDesc: `Desc{fqName: "test_metric", help: "Help string", constLabels: {}, variableLabels: [test_metric]}`,
+			oidToPdu:        make(map[string]gosnmp.SnmpPDU),
+			expectedMetrics: map[string]string{`label:<name:"test_metric" value:"-2" > gauge:<value:1 > `: `Desc{fqName: "test_metric", help: "Help string", constLabels: {}, variableLabels: [test_metric]}`},
 		},
 	}
 
-	for _, c := range cases {
-		m := pduToSample(c.indexOids, c.pdu, c.metric, c.oidToPdu)
-		metric := &io_prometheus_client.Metric{}
-		err := m.Write(metric)
-		if err != nil {
-			t.Fatalf("Error writing metric: %v", err)
-		}
-		if metric.String() != c.expectedMetric {
-			t.Fatalf("Unexpected metric: got %v, want %v", metric.String(), c.expectedMetric)
+	for i, c := range cases {
+		metrics := pduToSamples(c.indexOids, c.pdu, c.metric, c.oidToPdu)
+		if len(metrics) != len(c.expectedMetrics) {
+			t.Fatalf("Unexpected number of metrics returned for case %v: want %v, got %v", i, len(c.expectedMetrics), len(metrics))
 		}
-		if m.Desc().String() != c.expectedMetricDesc {
-			t.Fatalf("Unexpected metric description: got %v, want %v", m.Desc().String(), c.expectedMetricDesc)
+		metric := &io_prometheus_client.Metric{}
+		for _, m := range metrics {
+			err := m.Write(metric)
+			if err != nil {
+				t.Fatalf("Error writing metric: %v", err)
+			}
+			if _, ok := c.expectedMetrics[metric.String()]; !ok {
```

Comment:
If you're going to do things that way, you also need to check that the length of the expected metrics matches what you actually got.